llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -563,16 +569,37 @@ Instruction *emitSpecConstantRecursiveImpl(Type *Ty, Instruction *InsertBefore,
       IDs.push_back({IDs.back().ID + 1, true});
     }
     Elements.push_back(Def);
+    Index++;
   };
-  auto LoopIteration = [&](Type *Ty, unsigned LocalIndex) {
+  auto LoopIteration = [&](Type *ElTy, unsigned LocalIndex) {
     // Select corresponding element of the default value.
     // There are cases when provided default value contains less elements than
     // specialization constants: it could happen when a struct is extended with
     // a padding to make its size aligned. In such cases, we simply initialize
     // any "extra" elements with undef.
-    Constant *ElemDefaultValue = DefaultValue->getAggregateElement(LocalIndex);
-    if (!ElemDefaultValue)
-      ElemDefaultValue = UndefValue::get(Ty);
+    const auto ElemDefaultValue = [&]() -> Constant * {
+      if (auto *StructTy = dyn_cast<StructType>(Ty)) {
+        auto *DefaultValueType = dyn_cast<StructType>(DefaultValue->getType());
+        assert(DefaultValueType && "Default value is a struct type while"
+                                   "spec constant type is not!");
```

Comment:
```suggestion
        auto *DefaultValueType = cast<StructType>(DefaultValue->getType());
```

`dyn_cast` contains a runtime `if` to return `nullptr`. If you are not checking for `nullptr`, then a regular `cast` is enough and it already contains `assert(isa<T>)` within it